src/Build.OM.UnitTests/Definition/Project_Tests.cs

```diff
@@ -2866,7 +2866,7 @@ public void GetItemProvenanceResultsShouldBeInItemElementOrder()
             var sb = new StringBuilder();
             for (int i = 0; i < itemElements; i++)
             {
-                sb.AppendLine($"<i_{i} Include=\"a\"/>");
+                sb.Append("<i_").Append(i).AppendLine(" Include=\"a\"/>");
```

Comment:
Is the goal here specifically for memory improvement? I'm not particularly great at reading IL, but it [looks like](https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQyAMABMugHQAqApgB7ADcOyAzEagQMIEDeOBvRzyACwEAsgAoAlFx59ZAZWAAnMADsA5gCEArpBgVFBAM4AjAgF4CKigHcCC5eu279k+tlmzVwAmHME8bh58JiQAggAO4RQqMAAyqhRiACQARAA8YAD6nGAAvgQAkioAxhBaemYAOikAhtUA9AB8KRKBQUbGYZHRMGLpWS1dUTFiYBJDPfFWfYUlZRXVdSlNLW28uTi5QA) there's an allocation that happens in the `AppendLine()` format with the integer, whereas the `Append().Append().AppendLine()` format has more instructions but no `box` being called and therefore no allocation.

Reply:
The benchmarking I mentioned [above ](https://github.com/dotnet/msbuild/pull/5629#issuecomment-671649072) was for this specifically:
```     [Benchmark]
        public StringBuilder AtOnce()
        {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < NumStrings; i++)
            sb.AppendLine($"hi {i} bye {DateTime.Now.ToString()} done");
            return sb;
        }
        [Benchmark]
        public StringBuilder WithFormat()
        {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < NumStrings; i++)
                sb.AppendFormat("hi {0} bye {1} done", i, DateTime.Now.ToString());
            return sb;
        }
        [Benchmark]
        public StringBuilder InParts()
        {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < NumStrings; i++)
                sb.Append("hi ").Append(i).Append(" bye ").Append(DateTime.Now.ToString()).AppendLine(" done");
            return sb;
        }
```

And it was pretty decisive. It does allocate but not nearly as much.

Reply:
(Although I think it allocates on the stack?)

Reply:
In any case:
![image](https://user-images.githubusercontent.com/12969783/90187504-ba4d7180-dd6e-11ea-9f11-6d2025bd5aee.png)